client/consensus/src/aux_schema.rs

```diff
@@ -80,19 +79,27 @@ pub fn transaction_metadata_key(ethereum_transaction_hash: H256) -> Vec<u8> {
 pub fn load_transaction_metadata<B: AuxStore>(
 	backend: &B,
 	hash: H256,
-) -> ClientResult<Option<(H256, u32)>> {
+) -> ClientResult<Option<Vec<(H256, u32)>>> {
 	let key = transaction_metadata_key(hash);
 	load_decode(backend, &key)
 }
 
 /// Update Aux transaction metadata.
-pub fn write_transaction_metadata<F, R>(
+pub fn write_transaction_metadata<F, R, Backend: AuxStore>(
+	client: &Backend,
 	hash: H256,
 	metadata: (H256, u32),
 	write_aux: F,
 ) -> R where
 	F: FnOnce(&[(&[u8], &[u8])]) -> R,
 {
 	let key = transaction_metadata_key(hash);
-	write_aux(&[(&key, &metadata.encode())])
+
+	let mut data: Vec<(H256, u32)> = match load_decode(client, &key) {
+		Ok(Some(metadata)) => metadata,
+		_ => Vec::new(),
```

Comment:
I'd suggest we distinguish `Ok(None)` and `Err`.`Ok(None)` is the case where we should return `Vec::new()`, but for `Err` -- it means we encountered a value that is present but we don't know how to decode. The latter means an internal failure of the consensus engine, and I think we shouldn't proceed in that case.

The current function does not support error returning, so we need to change that. I think that'll be a fine solution right now, but I also suggest in the future we move to use rocksdb's key prefix fetching -- store the transaction key as `hash + block_hash`, and a fetch by key prefix `hash` will return all values with corresponding `block_hash`. In that way this function will again not be able to fail.

Reply:
I'll open an issue for it.